proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,351 @@
+
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+---
+
+#### Option (donut option)
+
+Don't implement setIcon() for dark mode, as it's used for dynamic icons anyway.
+
+* Pros: Easy to implement.
+* Cons: Developers would need to find their own ways of detecting dark mode.
+
+---
+
+#### Option (combined option)
+
+manifest.json
+```
+{
+  "icon_variants": {
+    "dark": {
+      "128": "128_dark.png"
+    },
+    "light": {
+      "128": "128_light.png"
+    }
+  },
+  "action": {
+    "default_icon": {
+      "128": "128_action.png"
+    },
+    "icon_variants": {
+      "dark": {
+        "128": "128_action_dark.png"
+      },
+      "light": {
+        "128": "128_action_light.png"
+      }
+    }
+  }
+}
+```
+
+action.setIcon()
+```
+action.setIcon({
+  variants: {
+    "dark": {
+      "128": "128_action_dark.png"
+    },
+    "light": {
+      "128": LightImageData
+    }
+  },
+});
+```
+
+Schema
+
+manifest.json
+```
+// `Mode` is dark or light (if either is used, both are expected).
+const Modes: string[] = ["dark", "light];
+const Mode = Modes[Math.floor(Math.random() * Modes.length())];
+
+// Primitive type declaration.
+type Size = string;
+type Path = string;
+
+// Map icon size to path.
+const IconSizeToPath: map<Size, Path>;
+
+// Map mode to a string path or a dictionary mapping icon size to a string path.
+const IconVariants: map<Mode, IconSizeToPath | Path>;
+
+{
+  ...
+  "icon_variants": IconVariants;
+  "action": {
+    ...
+    "icon_variants": IconVariants
+  },
+};
+```
+
+action.setIcon()
+```
+// ImageData is an interface representing canvas element pixels.
+const IconSizeToImageData: map<Size, Path>;
+const IconVariantsWithImageData:
+  map<Mode, IconSizeToPath | IconSizeToImageData | ImageData | Path>;
+const ImageDataDictionary: map<Size, ImageData>;
+
+action.setIcon(
+  ...
+  icon_variants?: IconVariantsWithImageData,
+);
+```
+
+* Pros: Easy for developers to use.
+* Cons: Complex to define and implement.
+
+
+---
+
+#### Option (flat option)
+
+manifest.json
+```
+"icon_variants": [{
+  "16": "16.png",
+  "32": "32.png",
+}, {
+  "16": "dark16.png",
+  "32": "dark32.png",
+  "color_scheme": "dark",
+}, {
+  "16": "light16.png",
+  "32": "light32.png",
+  "color_scheme": "light"
+}]
+```
+
+action.setIcon()
+```
+action.setIcon({
+  variants: [
+    {
+      "16": "16.png",
+      "32": "32.png"
+    }, {
+      "16": darkImageData16,
+      "32": darkImageData32,
+      "color_scheme": "dark"
+    }, {
+      "16": lightImageData16,
+      "32": lightImageData32,
+      "color_scheme": "light"
+    }
+  ]
+});
+```
+
+* Pros: Future resistant allowing for more keys such as density (e.g. 2dppx),
+purpose (e.g. monochrome), and etc.
+* Cons: New paradigm for icon definition, and unclear how to handle color_schemes.
+* Questions: Should `color_scheme` be used instead of `mode`?
```

Comment:
The term `color_scheme` is indeed preferred over `mode` as it is used in other places on the web. For example, in the [`color-scheme` css property](https://developer.mozilla.org/en-US/docs/Web/CSS/color-scheme) and in the [`prefers-color-scheme` media feature](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme).